### ISSUE TITLE
Serialising covariance of CompoundSpectralModel with a TemplateSpectralModel

### ISSUE THREAD

**Issue by AtreyeeS** [gammapy/gammapy#4124]
**Bug description**

The covariance matrix is incorrectly written for a CompoundSpectralModel which is a product of a TemplateSpectralModel and a NormSpectralModel

**To Reproduce**

Either serialise the `diffuse_iem` model in the Fermi-LAT tutorial, or

```
from gammapy.modeling.models import *
t = TemplateSpectralModel(energy=[1,3,9]*u.TeV, 
                        values=[20,30,40]*1 / (u.cm*u.cm*u.MeV*u.s))
p = PowerLawNormSpectralModel()
sk = SkyModel(spectral_model=t*p, name="xx")
Models([sk]).write("test.yaml", overwrite=True)
m1 = Models.read("test.yaml")
```

This leads to an error 

```
File ~/Gammapy-dev/gammapy/gammapy/modeling/models/core.py:362, in DatasetModels.covariance(self, covariance)
    359 @covariance.setter
    360 def covariance(self, covariance):
    361     self._check_covariance()
--> 362     self._covariance.data = covariance
    364     for model in self._models:
    365         subcovar = self._covariance.get_subcovariance(model.covariance.parameters)

File ~/Gammapy-dev/gammapy/gammapy/modeling/covariance.py:59, in Covariance.data(self, value)
     57 shape = (npars, npars)
     58 if value.shape != shape:
---> 59     raise ValueError(
     60         f"Invalid covariance shape: {value.shape}, expected {shape}"
     61     )
     63 self._data = value

ValueError: Invalid covariance shape: (13, 12), expected (13, 13)
```

This occurs because the `norm` parameter is present on both the `TemplateSpectralModel` and the normed model, and is incorrectly serialised during writing.

**Other information**

It would be good to sort this before `1.0`
Otherwise, Fermi datasets serialisation is difficult.

**Comment by adonath** [comment]
Just to clarify: I think the issue is not that both parameters are called `norm`, but that the norm parameter of the `TemplateSpectralModel` is not serialized.

**Comment by registerrier** [comment]
I think the `norm` parameter is useful because it provides at least one degree of freedom. One could assume that all template spectral models come with a normalization parameter, which is not the case at the moment.
Of course the behavior can be obtained multiplying by a `NormSpectralModel` as in the example above.

**Comment by adonath** [comment]
So you propose to add the `norm` parameter to the `NDTemplate` as well? I think if we freeze them by default it is fine, however we have to make sure they are serialized correctly...

**Comment by AtreyeeS** [comment]
Removing the `norm` parameter creates a problem in the estimators, because the `reference_model` is currently serialised as a `TemplateSpectralModel`.

> Just to clarify: I think the issue is not that both parameters are called norm, but that the norm parameter of the TemplateSpectralModel is not serialized.

The parameter seems to be serialised okay. It is in writing the covariance that the error occurs.

**Comment by adonath** [comment]
@AtreyeeS As far as I can tell the issue is:

```python3
from gammapy.modeling.models import TemplateSpectralModel
from astropy import units as u

t = TemplateSpectralModel(energy=[1,3,9]*u.TeV, 
                        values=[20,30,40]*1 / (u.cm*u.cm*u.MeV*u.s))

print(t.norm)
print(t.to_dict())
```
The dict returned by `to_dict()` does not include the norm parameter...this is a bug in any case...

**Comment by AtreyeeS** [comment]
> @AtreyeeS I see what you mean, so we use the `TemplateSpectralModel` as the only spectral model component in a `SkyModel`, which correctly fails, because there is no amplitude like parameter. Then we go for @registerrier's solution:
> 
> * Fix serialization of the nom for the `TemplateSpectralModel`
> * Add a norm to `TemplateNDSpectralModel` as well for consistency

Okay I will try to do this

**Comment by AtreyeeS** [comment]
There are two issues here:

1. The `write_covariance` replaces one of the norms in the table
https://github.com/gammapy/gammapy/blob/c5697fd2c91fb9728686f44d8a19bfbfd103e5be/gammapy/modeling/models/core.py#L577

2. The `parameters_unique_names` repeats the name when the same parameter name appears twice in one model, as is the case here.

To reproduce:
```
from gammapy.modeling.models import *
import astropy.units as u
t = TemplateSpectralModel(energy=[1,3,9]*u.TeV, 
                        values=[20,30,40]*1 / (u.cm*u.cm*u.MeV*u.s))
p = PowerLawNormSpectralModel()
sk = SkyModel(spectral_model=t*p, name="xx")
ms = Models([sk])
ms.parameters_unique_names
```

output is: 
```
['xx.spectral.norm',
 'xx.spectral.norm',
 'xx.spectral.tilt',
 'xx.spectral.reference']
```

Since we can't have a Table with the col name repeated, the solution is to either not use tables, or maybe change the parameter name? Please suggest @adonath @registerrier

**Comment by adonath** [comment]
Thanks a lot @AtreyeeS for investigating, I see now that the problem is also related to the duplicated parameter names.

I guess the cleanest solution would be to go back the proposal to remove the `norm` parameter from the `TemplateSpectralModel` again and implement the fix in https://github.com/gammapy/gammapy/blob/master/gammapy/estimators/points/core.py#L239

In general my expectation is that the norm parameter will move to the `SkyModel` in future, so it is maybe good to already go in this direction?

**Comment by adonath** [comment]
Alternatively we could rename the parameter for `TemplateSpectralModel`, however the problem will always appear again for the `CompoundSpectralModel` if there are duplicated parameter names...I think there is no simple fix for this. Just forbidding duplicated parameter names, does not really work because it makes the `CompoundSpectralModel` un-usable for many use cases...

The best fix might be to extend the `parameters_unique_names` to include another level for the `CompoundSpectralModel`, like:

```
['xx.spectral.model1.norm',
 'xx.spectral.model2.norm',
 'xx.spectral.model2.tilt',
 'xx.spectral.model2.reference']
```
Right now I'm not sure how simple that is...

**Comment by AtreyeeS** [comment]
> I guess the cleanest solution would be to go back the proposal to remove the `norm` parameter from the `TemplateSpectralModel` again and implement the fix in https://github.com/gammapy/gammapy/blob/master/gammapy/estimators/points/core.py#L239

This does not really help, as the problem also occurs for other compound models (eg: additive ones)

```
import operator
from astropy import units as u
import matplotlib.pyplot as plt
from gammapy.modeling.models import (
    CompoundSpectralModel,
    LogParabolaSpectralModel,
    Models,
    PowerLawSpectralModel,
    SkyModel,
)

energy_bounds = [0.1, 100] * u.TeV
pwl = PowerLawSpectralModel(
    index=2.0, amplitude="1e-12 cm-2 s-1 TeV-1", reference="1 TeV"
)
lp = LogParabolaSpectralModel(
    amplitude="1e-12 cm-2 s-1 TeV-1", reference="10 TeV", alpha=2.0, beta=1.0
)
model = CompoundSpectralModel(pwl, lp, operator.add)
m = SkyModel(model)
Models([m]).write("test.yaml", overwrite=True)
m1 = Models.read("test.yaml")
```

> and then we should probably forbid the case, where either model1 or model2 is again another CompoundSpectralModel...

Is this safe? It might be useful to have additive models with some correction from norm models, no?

**Comment by adonath** [comment]
> Is this safe? It might be useful to have additive models with some correction from norm models, no?

Yes, but handling the hierarchy adds additional complexity. If we'd like to handle arbitrary complex hierarchies we need a recursive solution to generate the unique parameter names...

**Comment by AtreyeeS** [comment]
It is not ideal for sure, but can we use `np.savetxt` to save the covariance matrix? I tried, and it works without error

**Comment by adonath** [comment]
@AtreyeeS Ah, this way we would just rely on the order of the parameters. Technically this probably even works...when plotting we still get the duplicated parameter names, but this is maybe not that bad...

**Comment by AtreyeeS** [comment]
Technically, it works (I verified). In fact, we just use the covariance as an array now 
https://github.com/gammapy/gammapy/blob/c5697fd2c91fb9728686f44d8a19bfbfd103e5be/gammapy/modeling/models/core.py#L557

This means we will have to support this format in the future, even when we have a good solution for duplicated names... If that sounds okay, I can prepare a PR

**Comment by adonath** [comment]
Sounds good to me @AtreyeeS, I think it's good pragmatic solution for now!

**Comment by registerrier** [comment]
We impose unique model names, so that model_name+parameter_name should provide an unambiguous name, no?

Is the solution here backward compatible?

**Comment by AtreyeeS** [comment]
> Is the solution here backward compatible?

No!

**Comment by adonath** [comment]
@AtreyeeS Why? Isn't the order of the parameters reproducible?

**Comment by AtreyeeS** [comment]
Yes, the order is the same. but the first column right now is a string.

**Comment by adonath** [comment]
> but the first column right now is a string.

But this we could keep?

**Comment by registerrier** [comment]
Another option would to modify https://github.com/gammapy/gammapy/blob/c5697fd2c91fb9728686f44d8a19bfbfd103e5be/gammapy/modeling/models/core.py#L579
and replace `name` by the index

**Comment by registerrier** [comment]
Apparently changing to this:
```
def write_covariance(self, filename, **kwargs):
    """Write covariance to file
    Parameters
    ----------
    filename : str
        Filename
    **kwargs : dict
        Keyword arguments passed to `~astropy.table.Table.write`
    """
    names = self.parameters_unique_names
    table = Table()
    table["Parameters"] = names

    for idx, name in enumerate(names):
        values = self.covariance.data[idx]
        table[str(idx)] = values

    table.write(make_path(filename), **kwargs)
```
might be working.

We don't need to have the name repeated on the lines and columns of the Table

**Comment by AtreyeeS** [comment]
Indeed this works! Thanks a lot @registerrier

**Comment by adonath** [comment]
Resolved in #4128.
